File: src/main/java/de/hhu/stups/alloy2b/translation/ParserUtil.java
package de.hhu.stups.alloy2b.translation;

import java.util.Arrays;
import java.util.stream.Collectors;

final class ParserUtil {
    private ParserUtil() {
        throw new AssertionError("Utility class");
    }

    /**
     * Replace ticks by underscores and use single quotes for identifiers since strings with capital letter first are
     * variables in Prolog. The arity is also added to the Prolog term.
     */
    static String sanitizeIdentifier(String identifier) {
        if ("this".equals(identifier)) {
            // field declarations of signature may be joined with 'this' leading to a universal quantification in B
            return identifier;
        }
        String[] split = identifier.replace("'", "_")
            .replace("{", "")
            .replace("}", "")
            .split("/");
        return Arrays.stream(split)
            .filter(s -> !"this".equals(s))
            .map(s -> "'" + s + "'")
            .collect(Collectors.joining());
    }
}

File: src/main/kotlin/de/hhu/stups/alloy2b/translation/ParserUtil.kt
package de.hhu.stups.alloy2b.translation

/**
 * Replace ticks by underscores and use single quotes for identifiers since strings with capital letter first are
 * variables in Prolog. The arity is also added to the Prolog term.
 */
fun sanitizeIdentifier(identifier: String): String {
    if (identifier == "this") {
        // field declarations of signature may be joined with 'this' leading to a universal quantification in B
        return identifier
    }
    return identifier.replace("'", "_").replace("{", "").replace("}", "")
            .split("/").asSequence().filter { it != "this" }
            .joinToString("") { "'$it'" }
}
